engine/crates/engine-v2/query-planning/src/cost_estimation.rs

```diff
@@ -0,0 +1,156 @@
+            })
+            .collect::<Vec<_>>();
+
+        tracing::debug!("Estimating the cost for {} nodes", nodes.len());
+
+        let mut updated_any_node = true;
+        let mut iter_count = 0;
+        loop {
+            if iter_count >= MAX_ITERATIONS {
+                tracing::warn!("Exceeded maximum number of cost estimation iterations: {MAX_ITERATIONS}");
+                break;
+            }
+            if !updated_any_node {
+                break;
+            }
+
+            updated_any_node = false;
+            iter_count += 1;
+
+            for &dependent_node in &nodes {
+                let mut requirements_cost = self
+                    .graph
+                    .edges_directed(dependent_node, Direction::Outgoing)
+                    .filter_map(|edge| match edge.weight() {
+                        Edge::Requires => Some(edge.target()),
+                        _ => None,
+                    })
+                    .flat_map(|required_node| {
+                        self.graph
+                            .edges_directed(required_node, Direction::Incoming)
+                            .filter_map(|edge| match edge.weight() {
+                                Edge::Resolves => Some(edge.source()),
+                                _ => None,
+                            })
+                            .map(|resolver_of_required_node| {
+                                let ancestor = self.find_common_ancestor(dependent_node, resolver_of_required_node);
+                                self.estimate_field_cost_from(ancestor, resolver_of_required_node)
+                            })
+                            .min()
+                    })
+                    .max()
+                    .unwrap_or(0);
+                let incoming_edge = self
+                    .graph
+                    .edges_directed(dependent_node, Direction::Incoming)
+                    .next()
+                    .expect("Invariant: Resolver always have a single incoming edge")
+                    .id();
+                match self.graph.edge_weight_mut(incoming_edge).unwrap() {
+                    Edge::Resolver(edge_cost) => {
+                        // resolver have an inherent cost of 1.
+                        requirements_cost += 1;
+                        if *edge_cost != requirements_cost {
+                            updated_any_node = true;
+                            *edge_cost = requirements_cost
+                        }
+                    }
+                    Edge::CanResolveField(edge_cost) => {
+                        if *edge_cost != requirements_cost {
+                            updated_any_node = true;
+                            *edge_cost = requirements_cost
+                        }
+                    }
+                    _ => (),
+                }
+            }
+        }
+    }
+
+    fn find_common_ancestor(&self, mut left: NodeIndex, mut right: NodeIndex) -> NodeIndex {
```

Comment:
What if these two functions would return an `Option<NodeIndex>` so we don't panic in weird cases. Maybe log the error? I'm just a bit worried one of these functions has a bug which leads to the panic...

Reply:
good point, should return an internal planning error or something, rather than outright panicking